include/gsl/string_span

```diff
@@ -150,14 +151,42 @@ inline span<const char, dynamic_extent> ensure_z(const char* const& sz, std::ptr
 
 inline span<wchar_t, dynamic_extent> ensure_z(wchar_t* const& sz, std::ptrdiff_t max)
 {
-    auto len = details::wstring_length(sz, max);
+    auto len = details::string_length(sz, max);
     Ensures(sz[len] == 0);
     return {sz, len};
 }
 
 inline span<const wchar_t, dynamic_extent> ensure_z(const wchar_t* const& sz, std::ptrdiff_t max)
 {
-    auto len = details::wstring_length(sz, max);
+    auto len = details::string_length(sz, max);
+    Ensures(sz[len] == 0);
+    return {sz, len};
+}
+
+inline span<char16_t, dynamic_extent> ensure_z(char16_t* const& sz, std::ptrdiff_t max)
+{
+    auto len = details::string_length(sz, max);
+    Ensures(sz[len] == 0);
+    return {sz, len};
+}
+
+inline span<const char16_t, dynamic_extent> ensure_z(const char16_t* const& sz, std::ptrdiff_t max)
+{
+    auto len = details::string_length(sz, max);
+    Ensures(sz[len] == 0);
+    return {sz, len};
+}
+
+inline span<char32_t, dynamic_extent> ensure_z(char32_t* const& sz, std::ptrdiff_t max)
+{
+    auto len = details::string_length(sz, max);
+    Ensures(sz[len] == 0);
+    return {sz, len};
+}
+
+inline span<const char32_t, dynamic_extent> ensure_z(const char32_t* const& sz, std::ptrdiff_t max)
```

Comment:
I'll confess I haven't looked at this code in a little while, but is there a reason we wouldn't just have a generic `ensure_z<CharT>()`? All these individual implementations appear to do the same thing and have no type-specific code.

Reply:
Yes that's the obvious thing to do. But I tried it and failed. For some reason I could not figure out how to get that working. From what I could tell it looked like `char16_t` and `char32_t` as template parameters didn't resolve the same way? I think I could get it working for `char16_t` but not `char32_t` or maybe the other way round? My template knowledge is lacking so  maybe someone with better template foo can fix that?

Reply:
At line 131 there is a template version of the function. It is followed by more specialized overloads for `char` types. I am not sure why that is. Are those char type overloads even necessary?

I am adding a new commit that simply deletes the char type overloads. I can't see that they do anything different from the main templated function. I also changed the template parameter from `T` to `CharT` to be inline with the other generic functions.

Reply:
As a side note is `ensure_z()` such a great name for this function? It feels more like a `make_span_from_zstring()` or something that suggests it creates a string_span?

Reply:
Thanks @galik, deleting that old code was a nice bit of cleanup work (and removed a TODO that named me!). The name was originally chosen to distinguish between `assume_z()`: where it's a "trust-me" statement from the programmer. I would rather not rename at this point.